### PR TITLE
Chore: Moves QueryField to @grafana/ui

### LINKED ISSUE #19626: QueryField: make it available for external plugins
In order to simplify adding explore logs feature support for external plugins, it makes sense to share `QueryField` component with external plugins. Possibly, by moving it into `grafana-ui` package. Related to #18057 as well.

### PR DESCRIPTION
**Pull request by kaydelaney**
**What this PR does / why we need it**:
Moves QueryField and associated plugins to @grafana/ui for external use.

**Which issue(s) this PR fixes**:
Closes #19626

### CODE REVIEW

File: packages/grafana-ui/src/components/Typeahead/Typeahead.tsx [hunk 1]
```diff
@@ -31,21 +30,28 @@ interface State {
...
 
-  constructor(props: Props) {
-    super(props);
-
-    const allItems = flattenGroupItems(props.groupedItems);
-    const longestLabel = calculateLongestLabel(allItems);
-    const { listWidth, listHeight, itemHeight } = calculateListSizes(props.theme, allItems, longestLabel);
-    this.state = { listWidth, listHeight, itemHeight, hoveredItem: null, typeaheadIndex: 1, allItems };
-  }
+  state: State = { hoveredItem: -1, typeaheadIndex: 1, allItems: [], listWidth: -1, listHeight: -1, itemHeight: -1 };
```
Comment (davkal) [inline packages/grafana-ui/src/components/Typeahead/Typeahead.tsx]: using -1 for an unset `hoveredItem` will not work with its use of conditional logic in `render`

File: packages/grafana-ui/src/slate-plugins/braces.test.tsx [hunk 2]
```diff
@@ -11,7 +11,7 @@ declare global {
 }
 
 describe('braces', () => {
-  const handler = BracesPlugin().onKeyDown;
+  const handler = BracesPlugin().onKeyDown!;
```
Comment (davkal) [inline packages/grafana-ui/src/slate-plugins/braces.test.tsx]: ❗️ 🚀

**Comment by davkal** [comment]
Great PR, thanks for making some of the code paths more robust. 

Minor thing with `hoveredItem`.

File: packages/grafana-ui/src/components/Typeahead/Typeahead.tsx (hunk 1 above)
Comment (kaydelaney) [reply packages/grafana-ui/src/components/Typeahead/Typeahead.tsx]: ah, thanks for that catch, will fix

File: public/app/plugins/datasource/loki/language_provider.ts [hunk 3]
```diff
@@ -6,12 +6,13 @@ import { parseSelector, labelRegexp, selectorRegexp } from 'app/plugins/datasour
 import syntax from './syntax';
 
 // Types
-import { CompletionItem, LanguageProvider, TypeaheadInput, TypeaheadOutput, HistoryItem } from 'app/types/explore';
+import { LanguageProvider, HistoryItem } from 'app/types/explore';
```
Comment (alexanderzobnin) [inline public/app/plugins/datasource/loki/language_provider.ts]: Is it possible to move `LanguageProvider` to grafana-ui as well? It should be helpful for external plugins when you implement your own language provider.

Comment (kaydelaney) [reply public/app/plugins/datasource/loki/language_provider.ts]: or grafana-data? Any thoughts on this @torkelo ?

**Review by davkal — DISMISSED** [review]
LGTM

File: public/app/plugins/datasource/loki/language_provider.ts (hunk 3 above)
Comment (torkelo) [reply public/app/plugins/datasource/loki/language_provider.ts]: just the interface?

``´
export abstract class LanguageProvider {
  datasource: DataSourceApi;
  request: (url: string, params?: any) => Promise<any>;
  /**
   * Returns startTask that resolves with a task list when main syntax is loaded.
   * Task list consists of secondary promises that load more detailed language features.
   */
  start: () => Promise<any[]>;
  startTask?: Promise<any[]>;
}
```

It cannot be in @grafana/data yet as DataSourceApi is in @grafana/ui

Comment (torkelo) [reply public/app/plugins/datasource/loki/language_provider.ts]: or do you mean extract a base implementation of the language provider?

**Review by davkal — APPROVED** [review]
Tested field in Prometheus, Loki, and data links.

This PR does not change the slate exports, so I think we can safely merge this.

**Review by torkelo — APPROVED** [review]
Great job! 

Can't find anything major that needs fixing now :)

**Pull request merged**